src/bind.c

```diff
@@ -113,8 +113,13 @@ static SEXP vec_rbind(SEXP xs,
     }
   }
 
+  SEXP proxy = PROTECT(vec_proxy(ptype));
+  if (!is_data_frame(proxy)) {
```

Comment:
Probably this check should be done in `vec_proxy()`?

Reply:
I'm not sure the best way to approach this right now. The check would have to be pretty low level, probably in `vec_proxy_invoke()`, with something like:

```c
if (is_data_frame(x) && !is_data_frame(proxy)) {
  error();
}
```

This also feels related to `vec_is_list()` where we currently validate that the proxy is also a list, so the check might not just be specific to data frames

Reply:
I think it is generally related to inheriting from a base type. For instance if you inherit from `"integer"` it should be illegal to return a different type from the proxy. We can figure this out later and I guess an extra check doesn't hurt.